src/script/interpreter.cpp

```diff
@@ -1643,29 +1690,33 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                     valtype& vchSig    = stacktop(-3);
                     valtype& vchData   = stacktop(-2);
                     valtype& vchPubKey = stacktop(-1);
+                    bool fSuccess;
+                    // Different semantics for CHECKSIGFROMSTACK for taproot and pre-taproot
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0)
+                    {
+                        // Sigs from stack have no hash byte ever
+                        if (!CheckSignatureEncoding(vchSig, (flags | SCRIPT_NO_SIGHASH_BYTE), serror) || !CheckPubKeyEncoding(vchPubKey, flags, sigversion, serror)) {
+                            //serror is set
+                            return false;
+                        }
 
-                    // Sigs from stack have no hash byte ever
-                    if (!CheckSignatureEncoding(vchSig, (flags | SCRIPT_NO_SIGHASH_BYTE), serror) || !CheckPubKeyEncoding(vchPubKey, flags, sigversion, serror)) {
-                        //serror is set
-                        return false;
-                    }
-
-                    valtype vchHash(32);
-                    CSHA256().Write(vchData.data(), vchData.size()).Finalize(vchHash.data());
-                    uint256 hash(vchHash);
-
-                    CPubKey pubkey(vchPubKey);
-                    bool fSuccess = pubkey.Verify(hash, vchSig);
+                        valtype vchHash(32);
+                        CSHA256().Write(vchData.data(), vchData.size()).Finalize(vchHash.data());
+                        uint256 hash(vchHash);
 
+                        CPubKey pubkey(vchPubKey);
+                        fSuccess = pubkey.Verify(hash, vchSig);
+                    } else {
+                        // New BIP 340 semantics for CHECKSIGFROMSTACK
+                        const XOnlyPubKey pubkey{uint256(vchPubKey)};
+                        if (!EvalTapScriptCheckSigFromStack(vchSig, pubkey, execdata, flags, vchData, sigversion, serror, fSuccess)) return false;
+                    }
                     popstack(stack);
                     popstack(stack);
                     popstack(stack);
                     stack.push_back(fSuccess ? vchTrue : vchFalse);
                     if (opcode == OP_CHECKSIGFROMSTACKVERIFY)
                         popstack(stack);
-
-                    if (!fSuccess)
-                        return set_error(serror, SCRIPT_ERR_CHECKSIGVERIFY);
```

Comment:
I think this line was not supposed to be deleted (or it was supposed to be moved)

Reply:
Good catch. I think this code always failed if provided with a non-empty sig `CHECKSIGSTACK`. Do we want to change the behavior so that we allow empty sigs to fail without aborting? Or stick with the existing behavior where CHECKSIGFROMSTACK cannot be failed. In which case, there is little difference between CHECKSIGFROMSTACK and the verify version.

Reply:
My thinking is that this was not intended in the original implementation and we allow CHECKSIGFROMSTACK to fail.

Reply:
We cannot fix this bug without a hardfork :)

Reply:
Yeah, I was intending to say that we should fix it for tapscript